File: smoke_tests/dart/log_injection.dart
// Log Injection vulnerabilities in Dart

import 'dart:io';
import 'package:logging/logging.dart';

final _logger = Logger('App');

// Test 1: Print with user input
void vulnerablePrint(String userInput) {
  // VULNERABLE: User input in print
  print('User action: $userInput');
}

// Test 2: Logger with user input
void vulnerableLoggerInfo(String userInput) {
  // VULNERABLE: User input in log
  _logger.info('Processing request: $userInput');
}

// Test 3: Logger warning with user data
void vulnerableLoggerWarning(String errorDetails) {
  // VULNERABLE: User-controlled error in warning
  _logger.warning('Error occurred: $errorDetails');
}

// Test 4: Logger severe with user data
void vulnerableLoggerSevere(String exceptionMessage) {
  // VULNERABLE: User input in severe log
  _logger.severe('Critical error: $exceptionMessage');
}

// Test 5: Debug print with user input
void vulnerableDebugPrint(String data) {
  // VULNERABLE: User input in debug log
  debugPrint('Debug data: $data');
}

// Test 6: stdout.write with user input
void vulnerableStdoutWrite(String message) {
  // VULNERABLE: User input to stdout
  stdout.write('Message: $message\n');
}

// Test 7: stderr.write with user input
void vulnerableStderrWrite(String error) {
  // VULNERABLE: User input to stderr
  stderr.write('Error: $error\n');
}

// Test 8: File logging with user input
Future<void> vulnerableFileLog(String logPath, String message) async {
  // VULNERABLE: User input in file log
  var file = File(logPath);
  await file.writeAsString(
    '${DateTime.now()}: $message\n',
    mode: FileMode.append,
  );
}

// Test 9: Multi-line log injection
void vulnerableMultilineLog(String userInput) {
  // VULNERABLE: Can inject fake log entries
  print('[INFO] User submitted: $userInput');
}

// Test 10: JSON log with user input
void vulnerableJsonLog(String username, String action) {
  // VULNERABLE: JSON injection in logs
  print('{"user": "$username", "action": "$action", "timestamp": "${DateTime.now()}"}');
}

// Test 11: Log with HTTP headers
void vulnerableHeaderLog(Map<String, String> headers) {
  // VULNERABLE: Headers in logs
  print('Request headers: $headers');
}

// Test 12: Log with query parameters
void vulnerableQueryLog(Map<String, String> queryParams) {
  // VULNERABLE: Query params in logs
  print('Query parameters: $queryParams');
}

// Test 13: Exception logging with user message
void vulnerableExceptionLog(String userMessage) {
  // VULNERABLE: User message in exception log
  try {
    throw Exception(userMessage);
  } catch (e) {
    print('Exception: $e');
  }
}

// Test 14: Audit log with user data
void vulnerableAuditLog(String userId, String action, String details) {
  // VULNERABLE: User-controlled audit log
  _logger.info('AUDIT: User $userId performed $action: $details');
}

// Test 15: Stack trace with user input
void vulnerableStackTraceLog(String errorContext) {
  // VULNERABLE: User context in stack trace log
  try {
    throw Exception('Error in: $errorContext');
  } catch (e, stackTrace) {
    print('$e\n$stackTrace');
  }
}

// Test 16: CSV log format
Future<void> vulnerableCsvLog(String logPath, String field1, String field2) async {
  // VULNERABLE: CSV injection in logs
  var file = File(logPath);
  await file.writeAsString(
    '$field1,$field2,${DateTime.now()}\n',
    mode: FileMode.append,
  );
}

// Test 17: Syslog-style logging
void vulnerableSyslog(String facility, String message) {
  // VULNERABLE: User input in syslog format
  print('<$facility> ${DateTime.now()}: $message');
}

// Test 18: Analytics logging
void vulnerableAnalyticsLog(String eventName, Map<String, dynamic> params) {
  // VULNERABLE: User-controlled analytics
  print('Analytics: $eventName - $params');
}

File: smoke_tests/swift/log_injection.swift
// Log Injection vulnerabilities in Swift
import Foundation
import os.log

class LogInjectionVulnerabilities {

    let logger = Logger(subsystem: "com.app", category: "security")

    // Test 1: print() with user input
    func logUserAction(username: String) {
        // VULNERABLE: User input in log
        print("User action: \(username)")
    }

    // Test 2: NSLog with user data
    func logLogin(email: String) {
        // VULNERABLE: Email could contain log injection
        NSLog("Login attempt for: %@", email)
    }

    // Test 3: os_log with user input
    func logRequest(path: String) {
        // VULNERABLE: Path from user
        os_log("Request path: %{public}@", path)
    }

    // Test 4: Logger API with user data
    func logActivity(userId: String, action: String) {
        // VULNERABLE: Both parameters from user
        logger.info("User \(userId) performed \(action)")
    }

    // Test 5: Error logging with stack trace manipulation
    func logError(userMessage: String) {
        // VULNERABLE: User can inject fake stack traces
        logger.error("Error occurred: \(userMessage)")
    }

    // Test 6: File logging
    func logToFile(message: String) {
        let logPath = "/var/log/app.log"
        let timestamp = ISO8601DateFormatter().string(from: Date())
        // VULNERABLE: Message injection
        let logLine = "\(timestamp) - \(message)\n"
        if let data = logLine.data(using: .utf8) {
            FileManager.default.createFile(atPath: logPath, contents: data)
        }
    }

    // Test 7: Debug logging with request data
    func logRequest(headers: [String: String]) {
        for (key, value) in headers {
            // VULNERABLE: Header values in logs
            print("Header: \(key) = \(value)")
        }
    }

    // Test 8: Audit log injection
    func auditLog(event: String, details: String) {
        let entry = [
            "timestamp": Date().description,
            "event": event,  // VULNERABLE
            "details": details  // VULNERABLE
        ]
        // Write to audit log
        print("AUDIT: \(entry)")
    }

    // Test 9: JSON in logs
    func logJsonPayload(payload: [String: Any]) {
        if let data = try? JSONSerialization.data(withJSONObject: payload),
           let str = String(data: data, encoding: .utf8) {
            // VULNERABLE: JSON payload could contain log injection
            logger.debug("Payload: \(str)")
        }
    }

    // Test 10: Multi-line log injection
    func logMultiLine(input: String) {
        // VULNERABLE: Newlines allow fake log entries
        logger.info("""
            Processing input:
            \(input)
            End of input
            """)
    }

    // Test 11: Format string vulnerability
    func logFormat(format: String, args: CVarArg...) {
        // VULNERABLE: User-controlled format string
        let message = String(format: format, arguments: args)
        NSLog("%@", message)
    }

    // Test 12: Sensitive data in logs
    func logTransaction(cardNumber: String, amount: Double) {
        // VULNERABLE: Sensitive data logged
        print("Transaction: card=\(cardNumber), amount=\(amount)")
        NSLog("Processing payment for card %@", cardNumber)
    }

    // Test 13: Exception message logging
    func logException(error: Error, context: String) {
        // VULNERABLE: Context from user
        logger.error("Exception in \(context): \(error.localizedDescription)")
    }
}
